pubsub.go

```diff
@@ -151,6 +151,7 @@ type PubSubRouter interface {
 type Message struct {
 	*pb.Message
 	ReceivedFrom peer.ID
+	VaidatorData interface{}
```

Comment:
@vyzo where is this field set, did you forget to add a file?

Reply:
Also, is this just a convenience thing since the data is already available by emitting data from the validators?

Reply:
Yes, this is for validators to set, and for subscribers to consume. Basically, I'd like to avoid having to unmarshal things twice

Reply:
it is initialized when we construct the message; it's for use from validators, they can set it to pass data to the application (eg. the unmarshalled payload to avoid unmarshalling twice).
This is a filecoin ask.

Reply:
@vyzo Ok, no problem with convenience functions.

Although, I'd like to point out that in less than 30 minutes you've just passed a change that causes problems if there are multiple validators for a single message since they may both clobber the `ValidateData` field. There is also no documentation for this behavior.

Reply:
Welcome to agile software engineering. It's up to the validators to coordinate, no way to enforce that.

Reply:
That is literally the opposite of what was agreed at https://github.com/libp2p/go-libp2p-pubsub/issues/198

Reply:
So what exactly would you propose instead of the envelope field?
Also, that's something we _could_ change when we implement #198 stuff.

Reply:
Also note that the application controls the validators; if it needs to attach overlapping validators (for messages that propagate across multiple topics), it can add coordination logic of its own in the validators.

Reply:
> So what exactly would you propose instead of the envelope field?

One of:
1. The validators can internally emit data (e.g. they put data in a `map[msgID]ValidatorData` where msgID can be a pointer or some other identifier
2. ValidateData can be map[topic]interface{} (we've already restricted to one validator per topic so this shouldn't run into any issues)

Given that this was asked for by @whyrusleeping I assume option 1 was problematic/annoying (although understanding why would be really helpful), so let's try option 2?

Reply:
you'll need a mutex too, and I don't want to bloat this thing further.

Reply:
This completely breaks the abstraction of pubsub as a _shared service_. There's no reason we can't have validators _return_ additional data (along with an error) and have the validation subsystem collect all the extra validation data.

Reply:
So what's the proper issue of this field?

Reply:
It allows the validation to pass data to the application; one common usage (eg in lotus) is to pass the decoded message to avoid parsing and decoding twice.

Reply:
I get that and that’s my intent as well , just seems like there was an extended convo about it and wondering if any issues or assumptions need to be held when using it this way

Reply:
The issue in the convo was about multiple validators competing for the coveted field, but it's a very rare occurrence to have a message that validates in multiple topics.